### PR TITLE
Spherical inclusions

### PR DESCRIPTION
**Pull request by arjunramaswami**
Implemented an example to simulate the response of a matrix material with spherical inclusions in which the FFT operations can also be offloaded to the GPU with the help of CuPy

### CODE REVIEW

**Comment by tdegeus** [comment]
Amazing @arjunramaswami and @ahenkes1 thank a lot and very interesting. I'll review asap!

File: small-strain/spherical-inclusions/linear_elasticity_even.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
+import numpy as np
+import scipy.sparse.linalg as sp
+import itertools
+
+# Activate (1) GPU computation or not (0)
+GPU = 1
+if GPU == 1:
+    import cupy as cp
```
Comment (tdegeus) [inline small-strain/spherical-inclusions/linear_elasticity_even.py]:
```suggestion
# True: use GPU computations. False: regular CPU computations
GPU = True
if GPU == False:
    import cupy as cp
```

File: small-strain/spherical-inclusions/linear_elasticity_even.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    distance = np.sqrt(
+        (coords[0] - center[0]) ** 2 + (coords[1] - center[1]) ** 2 + (
+                coords[2] - center[2]) ** 2)
+    return 1 * (distance <= radius)
+# ----------------------------------- GRID ------------------------------------
+
+ndim   = 3            # number of dimensions
+N      = 32           # number of voxels (assumed equal for all directions)
+ndof   = ndim**2*N**3 # number of degrees-of-freedom
+shape  = [N,N,N]  # number of voxels as list: [Nx,Ny,Nz]
```
Comment (tdegeus) [inline small-strain/spherical-inclusions/linear_elasticity_even.py]:
```suggestion
ndof   = ndim**2*N**3 # number of degrees-of-freedom
shape  = [N,N,N]      # number of voxels as list: [Nx,Ny,Nz]
```

File: small-strain/spherical-inclusions/linear_elasticity_even.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+norm[0,:,:] = 0.
+norm[:,0,:] = 0.
+norm[:,:,0] = 0.
+# - set projection operator                                   [grid of tensors]
+for i, j, l, m in itertools.product(range(3), repeat=4):
+    Ghat4[i,j,l,m] = norm*delta(i,m)*q[j]*q[l]
+
+# Activate (1) GPU computation or not (0)
+GPU = 1
+if GPU == 1:
```
Comment (tdegeus) [inline small-strain/spherical-inclusions/linear_elasticity_even.py]:
```suggestion
if GPU:
```

File: small-strain/spherical-inclusions/linear_elasticity_even.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    ifft = lambda x: np.fft.fftshift(np.fft.ifftn(np.fft.ifftshift(x),[N,N,N]))
+
+# functions for the projection 'G', and the product 'G : K : eps'
+G        = lambda A2   : np.real( ifft( ddot42(Ghat4,fft(A2)) ) ).reshape(-1)
+K_deps   = lambda depsm: ddot42(K4,depsm.reshape(ndim,ndim,N,N,N))
+G_K_deps = lambda depsm: G(K_deps(depsm))
+
+# ------------------- PROBLEM DEFINITION / CONSTITIVE MODEL -------------------
+
+# phase indicator: spherical inclusions of volume fraction cf
```
Comment (tdegeus) [inline small-strain/spherical-inclusions/linear_elasticity_even.py]:
```suggestion
# phase indicator: "nspheres" random spherical inclusions, with total volume fraction "cf"
```

File: small-strain/spherical-inclusions/linear_elasticity_even.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+phase = np.zeros((N, N, N), dtype=np.int)
+real_cf = np.array([phase[phase == 1].size / phase.size])
```
Comment (tdegeus) [inline small-strain/spherical-inclusions/linear_elasticity_even.py]: That makes no sense here, `phase` just contains zeros

File: small-strain/spherical-inclusions/linear_elasticity_even.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+cf = 0.25
+no_of_spheres = 0
```
Comment (tdegeus) [inline small-strain/spherical-inclusions/linear_elasticity_even.py]:
```suggestion
nspheres = 0
```

File: small-strain/spherical-inclusions/linear_elasticity_even.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
+import numpy as np
+import scipy.sparse.linalg as sp
+import itertools
+
+# Activate (1) GPU computation or not (0)
+GPU = 1
+if GPU == 1:
+    import cupy as cp
+
+def create_bin_sphere(matrix_size, center, radius):
```
Comment (tdegeus) [inline small-strain/spherical-inclusions/linear_elasticity_even.py:11]:
```suggestion
# function to create a sphere (defined by "center" and "radius") 
# as ones in a grid (defined by "shape") of zeros
def create_bin_sphere(shape, center, radius):
```

File: small-strain/spherical-inclusions/linear_elasticity_even.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+phase = np.zeros((N, N, N), dtype=np.int)
+real_cf = np.array([phase[phase == 1].size / phase.size])
+cf = 0.25
+no_of_spheres = 0
+max_r = 0
+min_r = N
+print("\n\n+--------------------------------------------+")
+print("| %-26s %15s |" % (
+    'Number of inclusions', 'Volume fraction'))
+print("+--------------------------------------------+")
```
Comment (tdegeus) [inline small-strain/spherical-inclusions/linear_elasticity_even.py]:
```suggestion
```
Can be left up to the user, the important thing is to understand the code

File: small-strain/spherical-inclusions/linear_elasticity_even.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+while (phase[phase == 1].size / phase.size) <= cf:
+    r = np.random.uniform(low=1, high=10, size=None)
+    if r > max_r:
+        max_r = r
+    if r < min_r:
+        min_r = r
```
Comment (tdegeus) [inline small-strain/spherical-inclusions/linear_elasticity_even.py]: To be honest I'm not entirely sure what is going on here

File: small-strain/spherical-inclusions/linear_elasticity_even.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+                                      size=(3, 1))
+    sphere = create_bin_sphere(np.shape(phase), sphere_center, r)
+    if np.any(phase[sphere == 1]) == 1:
+        pass
+    else:
+        phase[sphere == 1] = 1
+        no_of_spheres += 1
+        real_cf = np.array([phase[phase == 1].size / phase.size])
+        print("| %-20d  %20.3f |" % (no_of_spheres, float(real_cf)))
+print("+--------------------------------------------+\n")
```
Comment (tdegeus) [inline small-strain/spherical-inclusions/linear_elasticity_even.py]:
```suggestion
```
(same as above)

**Review by tdegeus — CHANGES_REQUESTED** [review]
Thanks a lot for this example! It is really nice to see how easy it is to use GPUs, and I'm sure it will be appreciated by the community. Furthermore, it is very nice that you took the trouble to vectorize the definition of the projection operator (and to comment it well!).

I made some comments. Beyond them my only concern is the definition of the grid of inclusions. First, I did not understand every detail (see comment), but second, and most important, I think that it should be entirely moved to a function (on top like you did for `create_bin_sphere`). The reason say this is that it is then even easier to read the red-thread, and concern oneself with the problem specific details later. Furthermore, it gives a nice way to split the numerical value of the parameters from the implementation. 

Let me know!!

cc @vondrejc

**Comment by ahenkes1** [comment]
@tdegeus by the way, the vectorization of the projection operator was not done by us, I think credits go to @vondrejc  ;)

**Comment by arjunramaswami** [comment]
HI @tdegeus, are there some other changes required from us for the pull request?

**Comment by tdegeus** [comment]
Sorry @arjunramaswami . Thanks a lot for your contribution it is highly appreciated! Good luck with the research you do with the code!

**Pull request merged**